File: udp_server.go
package main

import (
	"bufio"
	"log"
	"net"
	"os"
	"sia/test/lib"
	"strconv"
	"time"
)

// sendData sends data over the UDP connection with a specific dataType.
func sendData(conn *net.UDPConn, dataType int, data []byte) {
	packet := append([]byte{byte(dataType)}, data...)
	_, err := conn.Write(packet)
	if err != nil {
		log.Fatalf("Failed to send data: %v", err)
	}
}

// RunTestUDPClient reads data from a file line by line and sends it over UDP at a specified frequency.
func main() {
	filePath := "sample-signal-100hz.txt"
	frequencyHz := 100

	lib.InitEnvVars()

	lib.Print(lib.UDP_SERVICE, "Starting UDP client")

	serverAddr := net.UDPAddr{
		Port: int(lib.UDP_PORT),
		IP:   net.ParseIP(lib.UDP_ADDR),
	}
	conn, err := net.DialUDP("udp", nil, &serverAddr)
	if err != nil {
		lib.Print(lib.UDP_SERVICE, "Failed to connect to server")
		return
	}
	defer conn.Close()

	// Calculate the interval between sends based on the frequency
	interval := time.Second / time.Duration(frequencyHz)

	// Open the file for streaming
	file, err := os.Open(filePath)
	if err != nil {
		log.Fatalf("Failed to open file: %v", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		// Read each line as data
		line := scanner.Text()
		value, err := strconv.ParseFloat(line, 64)
		if err != nil {
			log.Printf("Failed to convert line to integer: %v", err)
			continue
		}

		bytes := lib.Float64ToBytes(value)
		// Send data
		sendData(conn, 0, bytes)
		lib.Print(lib.UDP_SERVICE, "Sent EKG sensor data:", value)

		// Wait for the next interval
		time.Sleep(interval)
	}

	sendData(conn, 3, []byte{9})

	if err := scanner.Err(); err != nil {
		log.Fatalf("Error reading file: %v", err)
	}
}

File: udp_server.py
import socket
import time

# Configuration
FILE_PATH = "2-sample-signal-100hz.txt"
UDP_ADDR = "192.168.1.100"  # Change to your server's IP
UDP_PORT = 5005  # Change to your server's port
FREQUENCY_HZ = 100
INTERVAL = 1.0 / FREQUENCY_HZ


# Convert float to bytes (MicroPython doesn't have struct module by default)
def float_to_bytes(value):
    return bytearray(str(value), "utf-8")


# Initialize UDP socket
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

try:
    with open(FILE_PATH, "r") as file:
        for line in file:
            try:
                value = float(line.strip())
                data = float_to_bytes(value)
                packet = bytearray([0]) + data  # Prepend data type byte
                sock.sendto(packet, (UDP_ADDR, UDP_PORT))
                print("Sent EKG sensor data:", value)
            except ValueError:
                print("Invalid line, skipping:", line.strip())

            time.sleep(INTERVAL)

    # Send termination packet
    sock.sendto(bytearray([3, 9]), (UDP_ADDR, UDP_PORT))
    print("Sent termination signal")
except OSError as e:
    print("Error accessing file:", e)
finally:
    sock.close()
